### DISCUSSION TITLE
Performance of reading parquet from S3 with LIMIT

### DISCUSSION THREAD

**Discussion by JohnsonKuan** [duckdb/duckdb#5012] [Q&A] [answered]
Hi,

I am using the HTTPFS extension to query parquet files in S3.

I observed that a query with LIMIT can take a long time to run whereas a GROUP BY is fast. I tested this with a parquet file that has 20M rows and 40 columns.

I was wondering if anyone else has experienced this and can provide insights on why I am seeing this sort of query performance with HTTPFS.

Examples below.

```
INSTALL httpfs;
LOAD httpfs;
SET s3_region='us-west-2';
SET s3_access_key_id='<AWS access key id>';
SET s3_secret_access_key='<AWS secret access key>';
```

Below takes around 1 minute
```
SELECT * 
from 's3://bucket/file1.parquet'
LIMIT 5
;
```

Below takes a few seconds
```
SELECT
    col_1,
    col_2,
    col_3,
    COUNT(1) as cnt
FROM 's3://bucket/file1.parquet'
GROUP BY 1,2,3
;
```

**Comment by Mytherin** [comment, marked as answer]
The difference is in the number of columns you are selecting. In the first case all columns have to be read, whereas in the second case only three columns (`col_1`, `col_2` and `col_3`) have to be read. Depending on how the Parquet file is set-up, there is a limit (hah) to how much a `LIMIT` clause can be pushed down. Specifically the first row-group still needs to be fetched and decompressed. This might cause the first query to require significantly more data than the second.

**Reply by JohnsonKuan** [reply]
Thanks for the quick response. I'll need to figure out how to setup the Parquet file to address the limit :) to the `LIMIT`.
